docs/explanation/overlays.rst

```diff
@@ -0,0 +1,176 @@
+****************
+``OVERLAY`` Step
+****************
+
+It is sometimes necessary to modify the base filesystem when generating a
+parts-based payload. On these occasions, the ``OVERLAY`` step provides the
+means to do so.
+
+Overlay Parameters
+------------------
+
+A part has three parameters that can be used to adjust how the overlay step
+works: ``overlay-packages``, ``overlay-script`` and ``overlay-filter``.
+``overlay-packages`` and ``overlay`` (the filter parameter) behave much the
+same way as the related parameters on the ``STAGE`` step. ``overlay-script``
+likewise behaves similarly to ``override-stage``, including having access to
+the ``partsctl`` command.
+
+An example parts section with overlay parameters looks as follows:
+
+.. code-block:: yaml
+
+    parts:
+      part_with_overlay:
+        plugin: nil
+        overlay_packages:
+          - ed
+        overlay_script: |
+          rm -f ${CRAFT_OVERLAY}/usr/bin/vi ${CRAFT_OVERLAY}/usr/bin/vim*
+          rm -f ${CRAFT_OVERLAY}/usr/bin/emacs*
+          rm -f ${CRAFT_OVERLAY}/bin/nano
+        overlay:
+          - bin
+          - usr/bin
+
+After running this part, the overlay layer (and the final package) will only
+contain ed as an editor, with vi/vim, emacs, and nano all having been
+removed.
+
+.. _overlay_visibility:
+
+Overlay Visibility
+------------------
+
+By default, a part does not get access to the filesystem overlay. However,
+if a part provides any overlay parameters or depends on another part that
+provides overlay parameters, the location of the overlay is made available
+in the ``${CRAFT_OVERLAY}`` environment variable. For example:
+
+.. code-block:: yaml
+
+    parts:
+      no-overlay-access:
+        plugin: nil
+        override-build: |
+          echo "${CRAFT_OVERLAY}" > "${CRAFT_PART_INSTALL}/empty"
+      direct-overlay-access:
+        plugin: nil
+        overlay-script: |
+          echo "${CRAFT_OVERLAY}" > "${CRAFT_PART_INSTALL}/direct"
+      indirect-overlay-access:
+        plugin: nil
+        after: [direct-overlay-access]
+        override-build: |
+          echo "${CRAFT_OVERLAY}" > "${CRAFT_PART_INSTALL}/indirect"
+
+Because ``no-overlay-access`` has no access to the overlay directory, the
+``no-overlay-access`` step will fail, as the ``CRAFT_OVERLAY`` environment
+variable is unset. Removing that attempted access will make this project
+build.
+
+.. _overlay_layers:
+
+Layers
+------
+
+Each part has an overlay layer, which acts on a shared storage area in the
+processing order of parts. If the part doesn't specify any overlay parameters,
+this overlay is empty. The overlay's integrity is checked with a checksum
+defined by the following diagram:
+
+.. image:: /images/overlay_checksum.svg
+   :alt: Diagram for generating the overlay checksum.
+
+Each layer combines its parameters with the checksum for the layer below it
```

Comment:
This sentence is a little awkward with the many "checksums". Here's a suggestion: `Each layer's checksum is the combination of the layer's properties and the checksum of the layer below it.`

Reply:
That's a significant improvement. I slightly adjusted your suggestion and then improved the rest of the paragraph too.